[esp32, ethernet] Fix compilation errors when framework version set to `latest` or `5.4.1`

asergunov opened the pull request:
# What does this implement/fix?

By default we use latest release of pioarduino platform 53.3.13. But it supports only idf versions 5.3.x. Setting version to 5.4.x or latest causes frustrating compilation/cmake errors.

To fix this errors user should specify unreleased pioarduino platform version or platformio/espressif32. It's not clear from documentation or error messages.

This fix puts the internet wisdom to configuration script when it should be.

1. When version is not specified or supported by pioarduino platform 53.3.13 it will use as it was
2. When version is specified and not supported by pioarduino platform 53.3.13 but supported by platformio/espressif32@6.11.0 it will use platformio/espressif32@6.11.0 silently
3. When version is not supported by both platform releases, the user will be asked for platform as before
  a. If user specified pioarduino platform and version is known as supported it will silently work as it was
  b. If user specified pioarduino platform and version is not known it will ask user for release
  c.  If user specified platformio/espressif32 plaform it will be used
4. When version is set to `latest`  platformio/espressif32@6.11.0 and esp-idf 5.4.1 wil lbe used
5. When version is set to `dev` the `https://github.com/espressif/esp-idf.git` and `https://github.com/platformio/platform-espressif32.git` will be used

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- esphome/esphome-docs#<esphome-docs PR number goes here>

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:

### Reccommened will use pioarduino platform as it was
```yaml
esp32:
  framework:
    type: esp-idf
```

### latest will use esp-idf 5.4.1 with platformio/espressif32 6.11.0 as documented

```yaml
esp32:
  framework:
    type: esp-idf
    version: latest
```

### dev will use esp-idf and platformio/espressif32 from git repositories as documented

```yaml
# Example config.yaml
esp32:
  framework:
    type: esp-idf
    version: dev
```

### version not supported by pioarduino platform 53.3.13, by supported by platformio/espressif32 6.11.0 will use platformio/espressif32

```yaml
# Example config.yaml
esp32:
  framework:
    type: esp-idf
    version: 5.4.1
```

### if version is supported by will use pioarduino platform 53.3.13 it will be used as it was

```yaml
# Example config.yaml
esp32:
  framework:
    type: esp-idf
    version: 5.3.3
```

### is pioarduino specified in platform_version and version all works as before

```yaml
esp32:
  framework:
    type: esp-idf
    version: 5.4.1
    platform_version: "https://github.com/pioarduino/platform-espressif32.git#develop"
```

### if pioarduino specified in platform_version and version is not known user will be asked for `release` as it was

```yaml
esp32:
  framework:
    type: esp-idf
    version: 5.6.1
    # release: "20250510"
    platform_version: "https://github.com/pioarduino/platform-espressif32.git#develop"
```


### platform_version is not specified and version is not known

ESP-IDF 5.6.1 may be supported by platformio/espressif32; please specify 'platform_version'.

```yaml
esp32:
  framework:
    type: esp-idf
    version: 5.6.1
```


## Checklist:
  - [ ] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

asergunov committed:
Improved framework version selection

* figuring out idf platform version by idf version if not set
* for esp-idf dev version platform version should be manually specified
* don't use pioarduino for riscv variants

asergunov committed:
Merge remote-tracking branch 'upstream/dev' into esp32-configuration

asergunov committed:
blanks

asergunov committed:
lint

asergunov committed:
formatting

asergunov committed:
cleanup

asergunov committed:
cleanup

asergunov committed:
Change recommenced version to 5.3.1 because 5.3.2 is not available on https://registry.platformio.org/tools/platformio/framework-espidf/versions

asergunov committed:
Workaround of build error in ethernet component on esp-idf 5.3.1

asergunov committed:
formatting

asergunov committed:
limit ethernet workaround by idf version range

swoboda1337 commented:
This doesn't make sense "don't use pioarduino for riscv variants". Why not? Pioarduino supports them. 
BTW I think the intent is to switch entirely to pioarduino.

swoboda1337 commented:
> @swoboda1337 oh I see. Let me play a bit more with it. I just had build problems on esp32c6.
> 
> Is it for both Arduino and idf configurations? Do you know why? I just put comment to code se we all will know.

Arduino support has not been added for pioarduino yet but there is a PR https://github.com/esphome/esphome/pull/8604 and it does work

You can see 2025.5 is using pioarduino for all idf variants: https://github.com/esphome/esphome/blob/2025.5.2/esphome/components/esp32/__init__.py#L303

C6 works fine I have tried it many times but note 54.x. is broken, have to use 55.x or 53.x

asergunov committed:
Use pioarduino by default

asergunov committed:
Merge remote-tracking branch 'upstream/dev' into esp32-configuration

asergunov commented:
Setting 5.4.1 in version gives

```
INFO ESPHome 2025.6.0-dev
INFO Reading configuration can-sniffer.yaml...
WARNING The selected ESP-IDF framework version is not the recommended one. If there are connectivity or build issues please remove the manual version.
WARNING The selected ESP-IDF framework version is not the recommended one. If there are connectivity or build issues please remove the manual version.
WARNING GPIO9 is a strapping PIN and should only be used for I/O with care.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq.html#why-am-i-getting-a-warning-about-strapping-pins
INFO Generating C++ source...
INFO Updating https://github.com/espressif/esp-protocols.git@mdns-v1.8.2
INFO Compiling app...
Processing can-sniffer (board: esp32-c6-devkitm-1; framework: espidf; platform: https://github.com/pioarduino/platform-espressif32.git#53.03.13)
---------------------------------------------------------------------------------------------------------------------------------------
HARDWARE: ESP32C6 160MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.50401.0 (5.4.1) 
 - tool-cmake @ 3.30.2 
 - tool-esptoolpy @ 4.8.6 
 - tool-mklittlefs @ 3.2.0 
 - tool-ninja @ 1.7.1 
 - tool-riscv32-esp-elf-gdb @ 14.2.0+20240403 
 - tool-xtensa-esp-elf-gdb @ 14.2.0+20240403 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-riscv32-esp @ 13.2.0+20240530
Reading CMake configuration...
-- Found Git: /usr/bin/git (found version "2.43.0")
-- git rev-parse returned 'fatal: not a git repository (or any of the parent directories): .git'
-- The C compiler identification is GNU 13.2.0
-- The CXX compiler identification is GNU 13.2.0
-- The ASM compiler identification is GNU
-- Found assembler: ~/.platformio/packages/toolchain-riscv32-esp@13.2.0+20240530/bin/riscv32-esp-elf-gcc
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: ~/.platformio/packages/toolchain-riscv32-esp@13.2.0+20240530/bin/riscv32-esp-elf-gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: ~/.platformio/packages/toolchain-riscv32-esp@13.2.0+20240530/bin/riscv32-esp-elf-g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Building ESP-IDF components for target esp32c6
NOTICE: Current idf-component-manager default lock file version is 1.0.0, but found 2.0.0 in ./src/esphome/.esphome/build/can-sniffer/dependencies.lock. Recreating lock file with the current version.
Manifest files have changed, solving dependencies.
.Updating lock file at ./src/esphome/.esphome/build/can-sniffer/dependencies.lock
Processing 1 dependencies:
[1/1] idf (5.4.1)
-- Project sdkconfig file ./src/esphome/.esphome/build/can-sniffer/sdkconfig.can-sniffer
-- Configuring incomplete, errors occurred!

fatal: not a git repository (or any of the parent directories): .git
CMake Error at ~/.platformio/packages/framework-espidf/tools/cmake/tool_version_check.cmake:36 (message):
  

  Tool doesn't match supported version from list ['esp-14.2.0_20241119']:
  ~/.platformio/packages/toolchain-riscv32-esp@13.2.0+20240530/bin/riscv32-esp-elf-gcc


  Please try to run 'idf.py fullclean' to solve it.

Call Stack (most recent call first):
  ~/.platformio/packages/framework-espidf/components/esp_common/project_include.cmake:10 (check_expected_tool_version)
  ~/.platformio/packages/framework-espidf/tools/cmake/build.cmake:457 (include)
  ~/.platformio/packages/framework-espidf/tools/cmake/build.cmake:690 (__build_process_project_includes)
  ~/.platformio/packages/framework-espidf/tools/cmake/project.cmake:717 (idf_build_process)
  CMakeLists.txt:3 (project)
```

swoboda1337 commented:
FYI:

53.x.yz is meant for IDF 5.3.x and Arduino x.y.z
54.x.yz is meant for IDF 5.4.x and Arduino x.y.z
55.x.yz is meant for IDF 5.5.x and Arduino x.y.z

asergunov committed:
Change only latest and dev

- latest version changed to 5.4.1
- if latest version is used the latest idf platform 6.11.0 is used
- if dev version is used the platform-espressif32 from repo is used
- make core application build on latest esp-idf

asergunov committed:
Fix for unreleased pioarduino platforms
- commented out pioarduino idf versions which not supported by pioarduino platform
- using latest esp-idf platform for released idf versions

asergunov commented:
@swoboda1337 

> 
> FYI:
>
>53.0x.yz is meant for IDF 5.3 and Arduino x.y.z
>54.0x.yz is meant for IDF 5.4 and Arduino x.y.z
>55.0x.yz is meant for IDF 5.5 and Arduino x.y.z

But 54 and 55 are not released so 5.4, 5.5 idf versions are difenetly not supported by latest 53.3.13 we have by default.

I've commented them out in the list and using native platform in this case.

What do you think?

swoboda1337 commented:
BTW what do you mean by this: "Using native platformio/espressif32 for idf version not supported by pioarduino"

asergunov committed:
Keep old logic working

swoboda1337 commented:
> @swoboda1337 updated description

BTW if support for plaformio is dropped it would simplify things. Can see the first two messages in this channel: https://discord.com/channels/429907082951524364/1377029109758038087

swoboda1337 commented:
Also I already have a PR with the ethernet changes: https://github.com/esphome/esphome/pull/8457

swoboda1337 commented:
Ethernet already fixed in https://github.com/esphome/esphome/commit/e58baab563ed9a93dd3f55edfaac2b6b7be4af03

ESP32 platform stuff also out of date
